apps/test-app/frontend/src/components/tree-widget/TreeWidget.tsx

```diff
@@ -170,23 +182,73 @@ export function StatelessTreeWidget(props: Omit<Props, "rulesetId">) {
   const dataProvider = useMemo((): TreeDataProvider => {
     return async (node?: TreeNodeItem): Promise<TreeNodeItem[]> => {
       const parent: HierarchyNode | undefined = node ? (node as any).__internal : undefined;
+      let lastNonGroupingNodeId: string | undefined;
+      // Retrieve Id of last non grouping node
+      if (parent) {
+        let nodeKey = parent.key;
+        if (HierarchyNode.isGroupingNode(parent)) {
+          for (let i = parent.parentKeys.length - 1; i >= 0; --i) {
+            if (!HierarchyNodeKey.isGrouping(parent.parentKeys[i])) {
+              nodeKey = parent.parentKeys[i];
+              break;
+            }
+          }
+        }
+        if (!HierarchyNodeKey.isGrouping(nodeKey)) {
+          if (typeof nodeKey === "string") {
+            lastNonGroupingNodeId = nodeKey;
+          } else {
+            lastNonGroupingNodeId = nodeKey.instanceKeys[0].id;
+          }
+        }
+      }
+      // Get limit that is set for lastNonGroupingNodeId
+      const mapKeyValue = hierarchyLevelSizeLimit.get(lastNonGroupingNodeId);
+      let limit: undefined | number | "unbounded";
+      if (mapKeyValue === undefined) {
+        hierarchyLevelSizeLimit.set(lastNonGroupingNodeId, { limit });
+      } else {
+        limit = mapKeyValue.limit;
+      }
       try {
         if (modelsTreeHierarchyProvider) {
-          return (await modelsTreeHierarchyProvider.getNodes({ parentNode: parent })).map(parseTreeNodeItem);
+          return (await modelsTreeHierarchyProvider.getNodes({ parentNode: parent, hierarchyLevelSizeLimit: limit })).map(parseTreeNodeItem);
         }
         return [];
       } catch (e) {
+        if (e instanceof RowsLimitExceededError) {
+          return [
+            createInfoNode(
+              node,
+              `${IModelApp.localization.getLocalizedString("Sample:controls.result-limit-exceeded")} ${limit!}`,
+              InfoTreeNodeItemType.ResultSetTooLarge,
+            ),
+          ];
+        }
         // eslint-disable-next-line no-console
         console.error(e);
         return [];
       }
     };
-  }, [modelsTreeHierarchyProvider]);
+  }, [modelsTreeHierarchyProvider, hierarchyLevelSizeLimit]);
   const modelSource = useTreeModelSource(dataProvider);
   const nodeLoader = useTreeNodeLoader(dataProvider, modelSource);
   const eventHandler = useMemo(() => new TreeEventHandler({ nodeLoader, modelSource }), [nodeLoader, modelSource]);
   const treeModel = useTreeModel(modelSource);
-
+  const nodeRenderer = useCallback((nodeProps: TreeNodeRendererProps) => {
+    return (
+      <StatelessTreeNodeRenderer
+        {...nodeProps}
+        onLimitReset={(parentId?: string) =>
+          setHierarchyLevelSizeLimit((map) => {
+            const newMap = new Map(map);
+            newMap.set(parentId, { limit: "unbounded" });
+            return newMap;
+          })
```

Comment:
I think we could store those custom limits in a plain object rather than a `Map`, e.g. `useState<{[parentId: string]: number | "unbounded"}>({})`. Updating it here would be simpler too:
```ts
setHierarchyLevelSizeLimit((map) => ({ ...map, [parentId ?? ""]: "unbounded" }))
```

Reply:
Changed to your suggestion.